src/libcore/slice/mod.rs

```diff
@@ -3467,7 +3470,7 @@ impl<T> AsRef<[T]> for Iter<'_, T> {
 /// [slices]: ../../std/primitive.slice.html
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct IterMut<'a, T: 'a> {
-    ptr: *mut T,
+    ptr: NonNull<T>,
     end: *mut T, // If T is a ZST, this is actually ptr+len.  This encoding is picked so that
     // ptr == end is a quick test for the Iterator being empty, that works
     // for both ZST and non-ZST.
```

Comment:
Can they both be `NonNull<T>`? IIUC it's an invariant that `ptr <= end`, so a null end is invalid.

We don't use multiple niches yet, but a future optimization might, and it could help communicate the invariant better.

Reply:
According to a [comment further down](https://github.com/rust-lang/rust/blob/9a654c8f6ab724f06b1eae4d578a0b346b63a160/src/libcore/slice/mod.rs#L3246) end might be `0` in case of ZST.

Thanks for review, BTW! :)

Reply:
Ah of course: a ZST slice can have an arbitrary starting pointer, and len is encoded as the wrapping subtraction of end - start. Thus a slice of length 1 with the max-int pointer results in end = 0.

You know, now that we have unions a simpler way to do all this could be:

```
union LenOrEnd<T> {
    end: NonNull<T>,
    len: usize,
}

pub struct IterMut<'a, T> {
    marker: PhantomData<&'a mut T>,
    start: NonNull<T>,
    len_or_end: LenOrEnd<T>,
}
```

A union of `SizedIterMut` and `ZeroSizedIterMut` is another possibility. Except that currently union's don't participate in layout optimization. :/

Reply:
I'll bet you this will have slightly better performance too by representing len directly rather than needing to do arithmetic to derive it.

Reply:
I was thinking that the code might be improved in different ways too. I just think that we should go with smaller increments. The PR already improves something, new improvements could be in the next PR.

Anyway, I think the layout would have to be determined in type, not in union in order to enable optimizations. I'm not an expert but I suspect that making rustc smart enough to understand which part of the union is used is based on type and conditionally optimize layout would be very difficult.

Reply:
Yeah, I agree that an incremental improvement is fine right now.

The above does get the `NonNull<T>` layout optimization as the `NonNull<T>` is in the struct, rather than the union: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=82eb70c78825be8dd5e48fd5d8d521cc